main/exec.go

```diff
@@ -25,11 +26,74 @@ func Exec(ctx *log.Context, cmd, workdir string, stdout, stderr io.WriteCloser,
+
 	commandArgs, err := SetEnvironmentVariables(cfg)
-	cmd = cmd + commandArgs // Add command args if any
+	// Add command args if any. Unnamed arguments go in 'commandArgs'. Named arguments are set as environment variables so the'd be available within the script.
+	cmd = cmd + commandArgs
 
 	//executionMessage := ""   // TODO: return
 	exitCode := 0 // TODO: return exit code and execution state
+
+	if cfg.publicSettings.RunAsUser != "" {
+		ctx.Log("message", "RunAsUser is "+cfg.publicSettings.RunAsUser)
+
+		// Check prefix ("/var/lib/waagent/run-command-handler") exists in script path for ex. /var/lib/waagent/run-command-handler/download/<runcommandName>/0/script.sh
+		if !strings.HasPrefix(scriptPath, dataDir) {
+			errMessage := "Failed to determine RunAs script path. Contact ICM team AzureRT\\Extensions for this service error."
+			ctx.Log("message", errMessage)
+			return failedExitCodeGeneral, errors.New(errMessage)
+		}
+
+		// Gets suffix "download/<runcommandName>/0/script.sh"
+		downloadPathSuffix := scriptPath[len(dataDir):]
+		// formats into something like "/home/<RunAsUserName>/waagent/run-command-handler-runas/download/<runcommandName>/0/script.sh", This filepath doesn't exist yet.
+		runAsScriptFilePath := filepath.Join(fmt.Sprintf(runAsDir, cfg.publicSettings.RunAsUser), downloadPathSuffix)
+		runAsScriptDirectoryPath := filepath.Dir(runAsScriptFilePath) // Get directory of runAsScript that doesn't exist yet
+
+		// Create a 4-line script to be run as below to be able to Run As different user.
+		scriptLines := [4]string{}
+
+		// Create runAsScriptDirectoryPath and its intermediate directories if they do not exist
+		scriptLines[0] = fmt.Sprintf("mkdir -p -m u=rwx %s", runAsScriptDirectoryPath)
+
+		// Copy script at scriptPath to runAsScriptDirectoryPath
+		scriptLines[1] = fmt.Sprintf("cp %s %s", scriptPath, runAsScriptDirectoryPath)
+
+		// Provide read and execute permissions to RunAsUser on .sh file at runAsScriptFilePath
+		scriptLines[2] = fmt.Sprintf("chown -R %s %s", cfg.publicSettings.RunAsUser, runAsScriptDirectoryPath)
+
+		// echo pipes the RunAsPassword to sudo -S for RunAsUser instead of prompting the password interactively from user and blocking.
+		// echo <cfg.protectedSettings.RunAsPassword> | sudo -S -u <cfg.publicSettings.RunAsUser> <command>
+		scriptLines[3] = fmt.Sprintf("echo %s | sudo -S -u %s %s", cfg.protectedSettings.RunAsPassword, cfg.publicSettings.RunAsUser, runAsScriptFilePath+commandArgs)
+
+		// Create a shell script file that is run by root at <scriptPathDirectory>/runAsScript.sh that contains script that is run as RunAsUser
+		// ex. /var/lib/waagent/run-command-handler/download/<RunCommandName>/<sequenceNumber>/runAsScript.sh
+		// Create a writer
+		runAsScriptContainerScriptFilePath := filepath.Join(scriptPathDirectory, "runAsScript.sh")
+		runAsScriptContainerScript, runAsScriptContainerScriptCreateError := os.Create(runAsScriptContainerScriptFilePath)
```

Comment:
The current user, i.e. root will still own a newly created file. Invoke chown after the file is created to ensure that it is owned by runAsUser.

Reply:
This is correct. runAsScript.sh should be run by root. This is because runAsUser does not have permissions for the directory structure /var/lib/waagent/run-command-handler/download/<RunCommandName>/<sequenceNumber>/
which runAsScript.sh is in.

root@VivL1:/var/lib/waagent/run-command-handler/download/RA_1025_28/0# chown vickyl runAsScript.sh
root@VivL1:/var/lib/waagent/run-command-handler/download/RA_1025_28/0# ls -lrt runAsScript.sh
-r-xr-x--- 1 vickyl root 450 Oct 26 02:24 runAsScript.sh

root@VivL1:/var/lib/waagent/run-command-handler/download/RA_1025_28/0# echo Crpext12345. | sudo -S -u vickyl ./runAsScript.sh
sudo: unable to execute ./runAsScript.sh: Permission denied